addons/interaction/functions/fnc_takeLauncher.sqf

```diff
@@ -0,0 +1,46 @@
+#include "..\script_component.hpp"
+/*
+ * Author: Timi007. Cplhardcore
+ * Pass launcher to another unit.
+ *
+ * Arguments:
+ * 0: Unit that passes the launcher <OBJECT>
+ * 1: Unit to pass the launcher to <OBJECT>
+ * 2: launcher classname <STRING>
+ * 3: Play passing animation <BOOL> (default: true)
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [_player, _target, "launch_NLAW_F"] call ace_interaction_fnc_passLauncher
+ *
+ * Public: No
+ */
+
+params ["_player", "_target", "_launcher", ["_animate", true, [true]]];
+TRACE_4("Pass launcher params",_player,_target,_launcher,_animate);
+
+if (_launcher isEqualTo "") exitWith {ERROR("No launcher specified.")};
+private _hasLauncher = secondaryWeapon _player;
+if (_hasLauncher == "") exitWith {ERROR("Cannot add launcher to target due to lack of inventory space.")};
+
+private _CfgWeapons = configFile >> "CfgWeapons" >> _launcher;
+private _attachments = _target weaponAccessories _launcher;
+private _launcherAmmo = (currentMagazine _unit) select {
+    _x in getArray (configFile >> "CfgWeapons" >> _launcherClassname >> "magazines")
+};
+_target removeWeapon _launcher;
+{
+        _unit removeMagazine _x;
+    } forEach _launcherAmmo;
+
+_player addWeapon _launcher;
+{_player addSecondaryWeaponItem _x;} forEach _attachments;
+{_unit addMagazine _x;} forEach _launcherAmmo;
+
+if (_animate) then {[_player, "PutDown"] call EFUNC(common,doGesture)};
+
+private _playerName = [_player] call EFUNC(common,getName);
+private _displayName = getText (_cfgWeapons >> "displayName");
```

Comment:
I'm not sure if that `_x in getArray (...)` check is necessary, are there times where a launcher has a non-compatible magazine loaded?

`addMagazine` should probably use a CBA / ACE function that also checks if the magazine can be added, and drops on the ground if it can't be.

```suggestion
private _weaponConfig = configFile >> "CfgWeapons" >> _launcher;
private _attachments = _target weaponAccessories _launcher;
private _magazines = (currentMagazine _unit) select {
    _x in getArray (_weaponConfig >> "magazines")
};

_target removeWeapon _launcher;
{ _unit removeMagazine _x } forEach _magazines;

_player addWeapon _launcher;
{ _player addSecondaryWeaponItem _x } forEach _attachments;
{ _unit addMagazine _x } forEach _magazines;

if (_animate) then {[_player, "PutDown"] call EFUNC(common,doGesture)};

private _playerName = [_player] call EFUNC(common,getName);
private _displayName = getText (_weaponConfig>> "displayName");
```

Reply:
https://cbateam.github.io/CBA_A3/docs/files/common/fnc_addWeaponWithoutItems-sqf.html is a far better choice, handles removing magazines and adding them back in the same containers they were in, with exact ammo counts.

Reply:
> I'm not sure if that `_x in getArray (...)` check is necessary, are there times where a launcher has a non-compatible magazine loaded?
> 
> `addMagazine` should probably use a CBA / ACE function that also checks if the magazine can be added, and drops on the ground if it can't be.

Note: the config path in the line 31 of the suggested change should be saved into a var to avoid a potentially repeated lookup.

Reply:
I left that part as is for the same reason john commented later, that `select` statement doesn't account for compatible magazines added through magazine wells.

If I had to guess, CBA / ACE has some function that also takes this into account

Reply:
Don't use (getArray ... "magazines") because it doesn't support mag wells (which I think are used for some rpgs?)
just use `compatibleMagazines`

also need to think about 2nd muzzle (smaw)

Reply:
Current issue with handling multiple muzzles, it breaks disposables